tools/site_admin/manage_user_sessions.php

```diff
@@ -0,0 +1,109 @@
+
+$username = $_POST['username'] ?? ($_GET['username'] ?? null);
+$action = get_enumerated_param($_POST, 'action', null, ['logout'], true);
+
+$title = _("Manage User Sessions");
+output_header($title, NO_STATSBAR);
+
+echo "<h1>$title</h1>\n";
+echo "<p>" . _("This page shows how many separate sessions (logins) a user has and the option of logging them out.") . "</p>";
+
+show_username_form($username);
+
+if ($username) {
+    try {
+        $user = new User($username);
+    } catch (NonexistentUserException $exception) {
+        echo "<p class='error'>" . _("Invalid username") . "</p>";
+        exit;
+    }
+
+    echo "<hr>";
+    echo "<h2>$username ($user->real_name)</h2>";
+
+    if ($action == "logout") {
+        delete_sessions($username);
+    } else {
+        show_sessions_form($username);
+    }
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
+function show_username_form(?string $username): void
+{
+    echo "<form method='GET'>\n";
+    echo "<p>" . _("Username") . ": ";
+    echo "<input name='username' type='text' value='$username' size='20' autocapitalize='none' required>\n";
+    echo "</p>";
+    echo "<input type='submit' value='" . attr_safe(_("Look up this user")) . "'>\n";
+    echo "</form>\n";
+}
+
+// ------------------------------------------------
+
+function show_sessions_form(?string $username): void
+{
+    $sql = sprintf(
+        "
+        SELECT MD5(sid) as sid_hash, expiration
+        FROM sessions
+        WHERE username = '%s'
+        ORDER BY expiration
+        ",
+        DPDatabase::escape($username)
+    );
+    $result = DPDatabase::query($sql);
+    echo "<p>" . _("User has the following sessions. Sessions that have expired are no longer valid and will be automatically cleaned up by the system.") . "</p>";
+    echo "<table class='basic'>\n";
+    {
+        echo "<tr>\n";
+        echo "<th>" . _("SID Hash") . "</th>\n";
+        echo "<th>" . _("Expires") . "</th>\n";
+        echo "</tr>\n";
+    }
+    while ([$sid_hash, $expiration] = mysqli_fetch_row($result)) {
+        echo "<tr>";
+        echo "<td>$sid_hash</td>";
+        $notes = $expiration < time() ? _("(session has expired)") : "";
+        echo "<td>" . icu_date_template("long+time", $expiration) . " $notes</td>";
+        echo "</tr>";
+    }
+
+    echo "</table>\n";
+
+    echo "<form method='POST'>\n";
+    echo "<input type='hidden' name='username' value='$username'>\n";
+    echo "<input type='hidden' name='action' value='logout'>\n";
+    echo "<p><input type='submit' name='submit' value='" . attr_safe(_("Logout User")) . "'></p>\n";
+    echo "</form>\n";
+}
+
+// -----------------------------------------------------------------------------
+
+function delete_sessions(string $username): void
+{
+    $sql = sprintf(
+        "
+        DELETE FROM sessions
+        WHERE username = '%s'
+        ",
+        DPDatabase::escape($username)
+    );
+    DPDatabase::query($sql);
```

Comment:
Nit: Should we do error checking here? It should be a matter of checking if the result is `false`.

Reply:
Fair question and I don't think so. If the SQL ended up being invalid for some reason, the `query()` function will raise an exception (and log it). If there are no sessions this will still pass and just delete 0 rows. Pretty much all the code in the repo assumes we can talk to the database (ie: no transient network/socket failures) so that is a possible failure scenario we don't generally worry about.